### PR TITLE
Soaring improvements

### PR DESCRIPTION
**Pull request by samuelctabor**
Separate the vario functionality into a separate class from the soaring/thermalling functionality.

The logic and performance is unchanged.

If there is a better way of accessing the required parameters than passing them as arguments please let me know. I suppose these parameters could belong to the Variometer object rather than the SoaringController.

### CODE REVIEW

File: libraries/AP_Soaring/Variometer.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    //float dVdt = SpdHgt_Controller->get_VXdot();
+    //netto_rate = netto_rate + aspd*dVdt/GRAVITY_MSS;
+    //GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "%f %f %f %f\n",temp_netto,dVdt,netto_rate,barometer.get_altitude());
+    return netto_rate;
+}
+
+void Variometer::get_altitude_wrt_home(float *alt)
+{
+    _ahrs.get_relative_position_D_home(*alt);
+    *alt *= -1.0f;
```
Comment (magicrub) [inline libraries/AP_Soaring/Variometer.cpp]: small optimization here, you could use:
*alt = -*alt;

why can't this just return a float?

File: libraries/AP_Soaring/AP_Soaring.cpp [hunk 2]
```diff
@@ -171,20 +171,21 @@ bool SoaringController::check_thermal_criteria()
...
+    float alt = _vario.alt;
 
-    if (soar_active && (AP_HAL::micros64() - _thermal_start_time_us) > ((unsigned)min_thermal_s * 1e6) && thermalability < McCready(_alt)) {
-        GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "Thermal weak, recommend quitting: W %f R %f th %f alt %f Mc %f\n", (double)_ekf.X[0], (double)_ekf.X[1], (double)thermalability, (double)_alt, (double)McCready(_alt));
+    if (soar_active && (AP_HAL::micros64() - _thermal_start_time_us) > ((unsigned)min_thermal_s * 1e6) && thermalability < McCready(alt)) {
+        GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "Thermal weak, recommend quitting: W %f R %f th %f alt %f Mc %f\n", (double)_ekf.X[0], (double)_ekf.X[1], (double)thermalability, (double)alt, (double)McCready(alt));
         return true;
-    } else if (soar_active && (_alt>alt_max || _alt<alt_min)) {
-        GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_ALERT, "Out of allowable altitude range, beginning cruise. Alt = %f\n", (double)_alt);
+    } else if (soar_active && (alt>alt_max || alt<alt_min)) {
```
Comment (magicrub) [inline libraries/AP_Soaring/AP_Soaring.cpp]: is this supposed to be "_vario.alt"?

**Comment by tridge** [comment]
I think you were bitten by bad timing with a recent bug in master that caused travis to fail. Can you rebase and force push the branch so travis re-runs?

File: libraries/AP_Soaring/AP_Soaring.cpp (hunk 2 above)
Comment (khancyr) [inline libraries/AP_Soaring/AP_Soaring.cpp]: @magicrub alt = _varia.alt , see a little before !

File: libraries/AP_Soaring/Variometer.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    if (fabsf(alt - _last_alt) > 0.0001f) { // if no change in altitude then there will be no update of ekf buffer
+        // Both filtered total energy rates and unfiltered are computed for the thermal switching logic and the EKF
+        float aspd = 0;
+        float roll = _ahrs.roll;
+        if (!_ahrs.airspeed_estimate(&aspd)) {
+            aspd = _aparm.airspeed_cruise_cm / 100.0f;
+        }
+        _aspd_filt = ASPD_FILT * aspd + (1 - ASPD_FILT) * _aspd_filt;
+        float total_E = alt + 0.5 *_aspd_filt * _aspd_filt / GRAVITY_MSS;                                                  // Work out total energy
```
Comment (khancyr) [inline libraries/AP_Soaring/Variometer.cpp]: nicpick : 0.5f

**Comment by OXINARF** [comment]
@tridge It's not that. The error is at https://travis-ci.org/ArduPilot/ardupilot/jobs/226589867#L3535. By the way, Travis always runs the latest merge between master and the PR, so if a job is restarted it will be based on master branch. Semaphore doesn't do this though.

**Comment by tridge** [comment]
looks good, with just a minor error in a Log_Write call. I've pushed to master with that fixed.
Thanks Samuel!

**Pull request closed without merging**